spec/app/ics-020-fungible-token-transfer/README.md

```diff
@@ -45,15 +45,25 @@ interface FungibleTokenPacketData {
 }
 ```
 
+Whether the sending chain is the original source of funds, whether it is redeeming funds received over this channel, or whether it is relaying funds received from a 3rd chain is encoded in the `denomination` field. If the sending chain is the original source of funds, `denomination` must not contain the character `/`. If it contains `/`, then `denomination` must be of the form `{ics20Port}/{ics20Channel}/{remote_denom}`, where `ics20Port` and `ics20Channel` are the ics20 port and channel on the remote chain where the funds originated. If `ics20Port == packet.sourcePort` and `ics20Channel == packet.sourceChannel`, then the remote blockchain is returning/redeeming funds originally sent from this blockchain over this connection.
```

Comment:
Well written and thanks for cleaning up the spec. 

The only thing that confuses me is the usage of remote. There seems to be a reference to a send from a remote chain to this blockchain. The `remote_denom` confuses me since it sounds like a denom which originated on the remote chain, but in reality it is likely from this blockchain (or some other blockchain). `remote_denom` would only ever be on the remote chain if sent via localhost or the chain incorrectly mints fake IBC tokens

Does that make sense? Maybe I'm misinterpreting the meaning?

Reply:
You are right, this should be clarified. I think `{ics20Port}/{ics20Channel}/{denom}` should be used.

Note that there are two cases when this may be set.

1. I send `foo` from A -> B on channel `channel-7`. When B returns this token to me, it will declare it as `transfer/channel-7/foo` (and I will correctly map it to the one I sent before.
2. I send `bar` from C -> B on `channel-3`. Then I send this token from B -> A on `channel-7`. (I believe) this token will be sent as `transfer/channel-3/bar`

I am not sure how to express the second case and it doesn't seem to be explained clearly in the existing spec. I would be happy if you could extend this (maybe 2 paragraphs?) to ensure both (1) and (2) are covered, as well as the basic "no `/` in denom" case.

Reply:
Yea, I'd be happy to write up something tomorrow. I find this area of ics20 always confusing, so I'll do my best to provide a concise and clear explanation

Reply:
Even more confusing is:

1. Send `bar` from C -> B
2. Send `transfer/channel-3/bar` (?) from B -> A
3. Return this token A -> B as `transfer/channel-7/transfer/channel-3/bar` ??? which then unescrows `transfer/channel-3/bar` on B
4. Return this token B -> C as `transfer/channel-3/bar` which then unescrows `bar` on C.

This is my guess, I wouldn't know if this is correct until I run a relayer in verbose mode with this setup.

In general multi-hop and stripping off prefixes would be nice to clearly define (maybe make a PR on top of this one if it become a larger issue).

Reply:
Actually, this will not keep extending like I described above, but hash the value in each hop, right?
I am getting confused just thinking about this flow.

Reply:
checkout the my [suggestion](https://github.com/confio/ibc/pull/1). Happy to swap some of my wording for other terms if it would provide more clarity

Reply:
The general format I went for:
- define that tokens some sort of record/history
- define how this is represented
- define how this is updated

Reply:
Nice addition. I merged your changes.

I have a question about denom hashing: https://github.com/confio/ibc/pull/1#pullrequestreview-667707214

Maybe adding a line or two to clarify this (even saying it is allowed but optional in the implementations and unspecified by the spec)

Reply:
Do you think [this](https://github.com/cosmos/ibc/pull/576#discussion_r638822798) clarifies those questions

Reply:
Yeah, those changes should resolve this